subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1526,10 +1526,17 @@ static ssize_t send_tls(struct tls_context *ctx, const void *buf,
 			size_t len, int flags)
 {
 	int ret;
+	size_t sent = 0;
 
-	ret = mbedtls_ssl_write(&ctx->ssl, buf, len);
+	while (sent < len) {
```

Comment:
Socket `send()` API for streams does not guarantee that the entire buffer will be sent, it's ok to send only part of the data. Because of that, it's the application's responsibility to make sure that `send()` is called enough times to send the entire buffer. Therefore I don't see a reason to duplicate such functionality here.

Reply:
@rlubos Thank for quick review. Make sense. Indeed patching "write" function is improper way. Could you be kind to advice:
Having issue with MQTT iot connection. Trying to send a 71k payload, results in tls_write of 4k size, wand committing least bytes.
`ssize_t ztls_sendmsg_ctx(struct tls_context *ctx, const struct msghdr *msg,
			 int flags)
{
	ssize_t len;
	ssize_t ret;
	int i;

	len = 0;
	if (msg) {
		for (i = 0; i < msg->msg_iovlen; i++) {
			ret = ztls_sendto_ctx(ctx, msg->msg_iov[i].iov_base,
					      msg->msg_iov[i].iov_len, flags,
					      msg->msg_name, msg->msg_namelen);
			if (ret < 0) {
				return ret;
			}
			len += ret;
		}
	}

	return len;
}`
msg_iovlen is 71e3, first call to ztls_sendto_ctx returns 4k and loop is proceeding to next message. Would it be right to add a loop here. Ar caller must split his message to proper msg_iov array?

Reply:
Thanks for clarifying the issue.

Indeed, current implmentation of `sendmsg()` doesn't seem right, it should not proceed to the next chunk if it wasn't able to send the previous one completely. Otherwise, the stream would not be consistent.

I see two possible solutions to this. I think the most proper way would be to break the loop in case of partial send, and let the application update the `msghdr` structure and call `sendmsg()` again. But since modifying the `msghdr` seems much more complicated than simply moving the buffer pointer for regular `send()/sendto()`, and we don't seem to implement proper looping in MQTT for instance I think it could be accepted to loop within `ztls_sendmsg_ctx`, unless others object.

Reply:
@rlubos Updated `ztls_sendmsg_ctx` could you be so kind to take one more look?